test/serviceprofiles/serviceprofiles_test.go

```diff
@@ -0,0 +1,170 @@
+package serviceprofiles
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/linkerd/linkerd2/testutil"
+)
+
+var TestHelper *testutil.TestHelper
+
+type testCase struct {
+	args           []string
+	deployName     string
+	expectedRoutes []string
+	namespace      string
+	sourceName     string
+	spName         string
+}
+
+func TestMain(m *testing.M) {
+	TestHelper = testutil.NewTestHelper()
+	os.Exit(m.Run())
+}
+
+func TestServiceProfiles(t *testing.T) {
+
+	testNamespace := TestHelper.GetTestNamespace("serviceprofile-test")
+	out, _, err := TestHelper.LinkerdRun("inject", "testdata/tap_application.yaml")
+	if err != nil {
+		t.Fatalf("linkerd inject command failed\n%s", out)
+	}
+
+	out, err = TestHelper.KubectlApply(out, testNamespace)
+	if err != nil {
+		t.Fatalf("kubectl apply command failed\n%s", out)
+	}
+
+	// wait for deployments to start
+	for _, deploy := range []string{"t1", "t2", "t3", "gateway"} {
+		if err := TestHelper.CheckPods(testNamespace, deploy, 1); err != nil {
+			t.Error(err)
+		}
+
+		if err := TestHelper.CheckDeployment(testNamespace, deploy, 1); err != nil {
+			t.Error(fmt.Errorf("Error validating deployment [%s]:\n%s", deploy, err))
+		}
+	}
+
+	testCases := []testCase{
+		{
+			sourceName:     "tap",
+			namespace:      testNamespace,
+			deployName:     "deploy/t1",
+			spName:         "t1-svc",
+			expectedRoutes: []string{"POST /buoyantio.bb.TheService/theFunction", "[DEFAULT]"},
+		},
+		{
+			sourceName:     "open-api",
+			namespace:      testNamespace,
+			spName:         "t3-svc",
+			deployName:     "deploy/t3",
+			expectedRoutes: []string{"DELETE /testpath", "GET /testpath", "PATCH /testpath", "POST /testpath", "[DEFAULT]"},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.sourceName, func(t *testing.T) {
+			routes, err := getRoutes(tc.deployName, tc.namespace)
+			if err != nil {
+				t.Fatalf("routes command failed: %s\n", err)
+			}
+
+			initialExpectedRoutes := []string{"[DEFAULT]"}
+
+			if !assertExpectedRoutes(initialExpectedRoutes, routes) {
+				t.Fatalf("Expected routes to have prefixes:\n%s\nbut got:\n%s",
```

Comment:
This error message doesn't always reflect the result of the two assertions in the `assertExpectedRoutes()` function. See suggestion below on passing `t *testing.T` into `assertExpectedRoutes()`.